### PR TITLE
Prevent data loss with the backup zset

### PR DESCRIPTION
**Pull request by git-hulk**
There are mainly two changes in this PR:

- Add the backup zset to prevent data loss when consuming
- Move the tries from the timer set member to score

For the first change is intuitive since we pop from ready queue and add it back to the timer set is NOT atomic, so it may have the data loss when occurring error in the middle. BTW, to keep the PR easier to review, we won't implement job recovery here.

For the second change, the motivation is to let us remove job from the timer set and backup easier.

### CODE REVIEW

File: engine/redis_v2/timer.go [hunk 1]
```diff
@@ -13,17 +15,22 @@ local zset_key = KEYS[1]
...
-local expiredMembers = redis.call("ZRANGEBYSCORE", zset_key, 0, now, "LIMIT", 0, limit)
+local expiredMembers = redis.call("ZRANGEBYSCORE", zset_key, 0, min_score, "WITHSCORES", "LIMIT", 0, limit)
 
 if #expiredMembers == 0 then
 	return 0
 end
 
-for _,v in ipairs(expiredMembers) do
-	local ns, q, tries, job_id = struct.unpack("Hc0Hc0HHc0", v)
+local toBeRemovedMembers = {}
```
Comment (Napoleon0621) [inline engine/redis_v2/timer.go:27]: Is it necessary to add lines of comments for this block of code?

File: engine/redis_v2/queue_test.go [hunk 2]
```diff
@@ -6,9 +6,9 @@ import (
 	"testing"
 	"time"
 
-	"github.com/go-redis/redis/v8"
-
 	"github.com/bitleak/lmstfy/engine"
+	"github.com/go-redis/redis/v8"
```
Comment (Napoleon0621) [inline engine/redis_v2/queue_test.go:10]: Would you consider grouping the imports in order to be consistent with other files?

File: engine/redis_v2/timer.go [hunk 3]
```diff
@@ -79,12 +88,24 @@ func (t *Timer) Name() string {
...
+	return timestamp, tries
+}
+
 func (t *Timer) Add(namespace, queue, jobID string, delaySecond uint32, tries uint16) error {
 	metrics.timerAddJobs.WithLabelValues(t.redis.Name).Inc()
 	timestamp := time.Now().Unix() + int64(delaySecond)
 
-	// struct-pack the data in the format `Hc0Hc0HHc0`:
-	//   {namespace len}{namespace}{queue len}{queue}{tries}{jobID len}{jobID}
+	score := encodeScore(timestamp, tries)
```
Comment (Napoleon0621) [inline engine/redis_v2/timer.go:106]: 1. I would suggest adding an output format example for this function.
2. The comments below seem to be irrelevant.

File: engine/redis_v2/timer.go [hunk 4]
```diff
@@ -94,11 +115,32 @@ func (t *Timer) Add(namespace, queue, jobID string, delaySecond uint32, tries ui
...
-	copy(buf[2+namespaceLen+2+queueLen+2+2:], jobID)
+	binary.LittleEndian.PutUint16(buf[2+namespaceLen+2+queueLen:], uint16(jobIDLen))
+	copy(buf[2+namespaceLen+2+queueLen+2:], jobID)
+
+	err := t.redis.Conn.ZAdd(dummyCtx, t.Name(), &redis.Z{Score: score, Member: buf}).Err()
+	if err != nil {
+		return err
+	}
+
+	_ = t.removeFromBackup(namespace, queue, jobID)
```
Comment (Napoleon0621) [inline engine/redis_v2/timer.go:126]: Question: what's the purpose of this operation?

**Review by iwinux — APPROVED** [review]

File: engine/redis_v2/timer.go (hunk 1 above)
Comment (git-hulk) [reply engine/redis_v2/timer.go:27]: Yes, will add more comments for them to make it easier to understand

File: engine/redis_v2/queue_test.go (hunk 2 above)
Comment (git-hulk) [reply engine/redis_v2/queue_test.go:10]: For this point, I think we should group the imports order, but we don't require this before the CI integrate the import order check. Do you think this makes sense?

File: engine/redis_v2/timer.go (hunk 4 above)
Comment (git-hulk) [reply engine/redis_v2/timer.go:126]: We should remove the job from backup after success to add it back to the timer set. I will update some comments about why we can ignore the error here.

File: engine/redis_v2/queue_test.go (hunk 2 above)
Comment (Napoleon0621) [reply engine/redis_v2/queue_test.go:10]: sure

**Comment by git-hulk** [comment]
@Napoleon0621 Does this PR looks good to you?

**Review by Napoleon0621 — APPROVED** [review]

**Comment by Napoleon0621** [comment]
> 

LGTM
